src/wui/savegametable.cc

```diff
@@ -0,0 +1,261 @@
+	}
+	return result;
+}
+
+const std::string SavegameTable::find_game_type(const SavegameData& savegame) const {
+	switch (savegame.gametype) {
+	case GameController::GameType::kSingleplayer:
+		/** TRANSLATORS: "Single Player" entry in the Game Mode table column. */
+		/** TRANSLATORS: "Keep this to 6 letters maximum. */
+		/** TRANSLATORS: A tooltip will explain the abbreviation. */
+		/** TRANSLATORS: Make sure that this translation is consistent with the tooltip. */
+		return _("SP");
+	case GameController::GameType::kNetHost:
+		/** TRANSLATORS: "Multiplayer Host" entry in the Game Mode table column. */
+		/** TRANSLATORS: "Keep this to 2 letters maximum. */
+		/** TRANSLATORS: A tooltip will explain the abbreviation. */
+		/** TRANSLATORS: Make sure that this translation is consistent with the
+		        tooltip. */
+		/** TRANSLATORS: %1% is the number of players */
+		return (boost::format(_("H (%1%)")) % savegame.nrplayers).str();
+	case GameController::GameType::kNetClient:
+		/** TRANSLATORS: "Multiplayer" entry in the Game Mode table column. */
+		/** TRANSLATORS: "Keep this to 2 letters maximum. */
+		/** TRANSLATORS: A tooltip will explain the abbreviation. */
+		/** TRANSLATORS: Make sure that this translation is consistent with the
+		        tooltip. */
+		/** TRANSLATORS: %1% is the number of players */
+		return (boost::format(_("MP (%1%)")) % savegame.nrplayers).str();
+	case GameController::GameType::kReplay:
+		return "";
+	case GameController::GameType::kUndefined:
+		NEVER_HERE();
+	}
+	NEVER_HERE();
+}
+
+void SavegameTable::fill(const std::vector<SavegameData>& entries) {
+	clear();
+
+	for (size_t i = 0; i < entries.size(); ++i) {
+		const SavegameData& savegame = entries[i];
+		UI::Table<uintptr_t const>::EntryRecord& te = add(i);
+
+		if (savegame.errormessage.empty()) {
+			create_valid_entry(te, savegame);
+		} else {
+			create_error_entry(te, savegame);
+		}
+	}
+	sort();
+	layout();
+	focus();
+}
+
+void SavegameTable::create_error_entry(UI::Table<uintptr_t const>::EntryRecord& te,
+                                       const SavegameData& savegame) {
+	size_t last_column_index = number_of_columns() - 1;
+	for (size_t col_index = 0; col_index < last_column_index; col_index++) {
+		te.set_string(col_index, "");
+	}
+	te.set_string(
+	   last_column_index, (boost::format(_("Incompatible: %s")) % savegame.mapname).str());
+}
+
+void SavegameTable::create_directory_entry(UI::Table<const uintptr_t>::EntryRecord& te,
+                                           const SavegameData& savegame) {
+	size_t last_column_index = number_of_columns() - 1;
+	for (size_t col_index = 0; col_index < last_column_index; col_index++) {
+		te.set_string(col_index, "");
+	}
+	if (savegame.is_parent_directory()) {
+		te.set_picture(last_column_index, g_gr->images().get("images/ui_basic/ls_dir.png"),
+		               (boost::format("%s") % _("<parent>")).str());
```

Comment:
This should be identical to `MapData::parent_name()`, otherwise we'll have additional work for translators that can introduce inconsistencies in the translations.

Reply:
done